### DISCUSSION TITLE
override User or DefaultUser type

### DISCUSSION THREAD

**Discussion by aboyo** [nextauthjs/next-auth#8923] [Help] [answered]
### Summary

Hi all,

I am currently using next-auth to implement login functionality. I am using the CredentialsProvider along with my own API server to verify usernames and passwords. In the authorize function, I want to return my custom-defined user object.

I have added a `types/next-auth.d.ts` file in the project and declared the User type: (it just only has one property 'id')

<img width="781" alt="截圖 2023-10-23 下午5 57 48" src="https://github.com/nextauthjs/next-auth/assets/21969308/1a4ce28f-2b2e-4b50-9a63-fc1e6e054381">

but, In the authorize function of the CredentialsProvider, user data can still have additional properties:

<img width="1392" alt="截圖 2023-10-23 下午6 08 21" src="https://github.com/nextauthjs/next-auth/assets/21969308/17ad269f-587d-4871-bd90-896379c30962">

I have also tried declaring DefaultUser in the `types/next-auth.d.ts` file, but it didn't work.

How do I customize the User type~?

// I'm using next-auth:^4.24.3

**Comment by nktnet1** [comment, marked as answer]
For next-auth v4.24.10 and v5.0.0-beta.25, I'm using the following:
```ts
/* eslint-disable @typescript-eslint/no-unused-vars */
import { JWT } from "next-auth/jwt"
import NextAuth, { DefaultSession, User } from 'next-auth';

export interface CustomAuthUser extends User {
  id: string;
}

declare module 'next-auth' {
  interface Session {
    user?: CustomAuthUser & DefaultSession['user'];
  }
}

declare module 'next-auth/jwt' {
  interface JWT {
    user?: CustomAuthUser;
  }
}
```

Bear in mind that your versions of `@auth/drizzle-adapter` and `@auth/core`, if you're using them in your project could affect the above - e.g.
- https://github.com/jschuur/decent-nextjs-starter-template/issues/33
- https://github.com/nextauthjs/next-auth/issues/9571#issuecomment-2191513065

and if these can't be updated, you may need to change
```ts
declare module 'next-auth/jwt' {
```
to
```ts
declare module '@auth/core/jwt' {
```
